### ISSUE TITLE
Stop sinking when buffer exceed max_size

### ISSUE THREAD

**Issue by CasparChou** [timberio/vector#6741]
### Vector Version
0.11.1

### Vector Configuration File

https://gist.github.com/CasparChou/70234ddf69d18291f27f04df106f0c5d

### Debug Output

https://gist.github.com/CasparChou/70234ddf69d18291f27f04df106f0c5d

### Expected Behavior

Sink logs continuously

### Actual Behavior

Stop sink and block incoming data

### Example Data

K8s logs, Journald

### Additional Context

Kubernetes Version
```
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.15-gke.7800", GitCommit:"cef3156c566a1d1a4b23ee360a760f45bfbaaac1", GitTreeState:"clean", BuildDate:"2020-12-14T09:12:37Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
```

CPU Usage
![截圖 2021-03-12 上午11 44 53](https://user-images.githubusercontent.com/10682620/110890511-25721d00-832b-11eb-9ee1-e2a1edefde6b.png)

Memory Usage
![截圖 2021-03-12 上午11 45 04](https://user-images.githubusercontent.com/10682620/110890515-299e3a80-832b-11eb-9ad7-07d722b09579.png)

Network Receive/Transmit (traffic should be at least 3-4MB/s, not KB/s level, due to buffer block)
![截圖 2021-03-12 上午11 45 23](https://user-images.githubusercontent.com/10682620/110890527-2d31c180-832b-11eb-85ce-6c0bf21bacc9.png)
(meanwhile with fluentbit/fluentd)
![image](https://user-images.githubusercontent.com/10682620/110891106-46873d80-832c-11eb-8ccd-8f4be83b6344.png)

Storage (buffer usage)
![截圖 2021-03-12 上午11 45 31](https://user-images.githubusercontent.com/10682620/110890535-2f941b80-832b-11eb-91b0-1e27777696b1.png)

Connections from agent
![截圖 2021-03-12 上午11 45 56](https://user-images.githubusercontent.com/10682620/110890537-315ddf00-832b-11eb-9d8e-9e0dd0b604fc.png)

**Comment by CasparChou** [comment]
UPDATE & explain

AM 10:41 start two aggregator pod, then two pod was not sinking anything

PM 12:33:57
`vector-aggregator-0` start printing lots of warn logs said `vector::sources::util::tcp: Failed to read data from TCP source. error=bytes remaining on stream`
then vector api stuck, so pod been terminated by k8s liveness healthcheck

this is healthcheck part added by myself in helm chart
```
          {{- if .Values.livenessProbe.enabled }}
          livenessProbe:
            httpGet:
              path: /health
              port: 8686
              scheme: HTTP
            failureThreshold: {{ .Values.livenessProbe.failureThreshold }}
            initialDelaySeconds: {{ .Values.livenessProbe.initialDelaySeconds }}
            periodSeconds: {{ .Values.livenessProbe.periodSeconds }}
            successThreshold: {{ .Values.livenessProbe.successThreshold }}
            timeoutSeconds: {{ .Values.livenessProbe.timeoutSeconds }}
          {{- end }}
```

PM 12:34:27
after restarting pod, aggregator-0 seems starting to sink logs, but very very small size
aggregator-1 still sleeping


PM 13:35
aggregator-0 flush half of buffer (2GB)
aggregator-1 still sleeping

PM 14:30
update helm chart release
aggregator-0 flush all of buffer
both aggregator-0 and aggregator-1 restarted
after restart pod, aggregator-1 still sleeping

![image](https://user-images.githubusercontent.com/10682620/110904339-c5d43b80-8343-11eb-9b56-69c4df480998.png)

**Comment by CasparChou** [comment]
workaround:

let healthcheck to force restart pod

```
            livenessProbe:
-             httpGet:
-               path: /health
-               port: 8686
-               scheme: HTTP
+             exec:
+               command:
+               - bash
+               - -c
+               - |
+                 which curl || (apt-get -qq update && apt-get -qq install -y curl);
+                 (($(du -sb /vector-data-dir/ | awk '{print $1}') < 3000000000)) &&
+                 curl --fail -s 0:8686/health > /dev/null
              failureThreshold: 5
              initialDelaySeconds: 120
              periodSeconds: 6
              successThreshold: 1
              timeoutSeconds: 5
```

**Comment by ktff** [comment]
@CasparChou could you also provide disk I/O graph for the aggregators as they will be really useful to confirm/eliminate issues with the disk buffer.

**Comment by CasparChou** [comment]
aggregator-1, the one that not sinking at 10am to 2pm
![image](https://user-images.githubusercontent.com/10682620/111094656-b7219a80-8576-11eb-9894-7a8cd972076a.png)

**Comment by mikhno-s** [comment]
The similar problem.

vector.toml:
```toml
data_dir = "/var/lib/vector"
[api]
  enabled = true
[sources.internal_metrics]
  type = "internal_metrics"
[sinks.my_sink_id]
  type = "prometheus_exporter"
  inputs = ["internal_metrics"]
  address = "0.0.0.0:9598"

[sources.in_envoy_logs]
  type = "file"
  include = ["/var/log/envoy/ingress.log"]
  exclude = ["*gz"]
  glob_minimum_cooldown = 1000
  ignore_older = 86400
  max_read_bytes = 4096

  fingerprinting.strategy = "checksum"
  fingerprinting.fingerprint_bytes = 256
  fingerprinting.ignored_header_bytes = 0

[sinks.s3_logs]
  inputs = ["in_envoy_logs"]
  type   = "aws_s3"
  bucket = "some-bucket"
  region = "us-east-1"
  encoding.codec = "ndjson"
  key_prefix = "${DC}/${NODE_NAME}/%Y/%m/%d/%H/%M/%S/"
  compression = "gzip"
  content_encoding = "gzip"
  buffer.when_full = "block"
  buffer.type = "disk"
  buffer.max_size = 1004900000 # 1GB
  batch.max_bytes = 10000000 # 10 MB
  batch.timeout_secs = 30
```

I have 3GB mem limit and OOM kills vector in 30-60s after start.
But even during that time vector reads logs, but does not send anything. I see it, because vector runs with `-vvvv` 

The problem gone after deleting all files in /var/lib/vector/s3_logs_buffer.

```shell
gke-ae1-app-8e09496f-z8gu /var/lib/vector/s3_logs_buffer # du -sh; ls -l |wc -l
4.7G	.
1312
```
   
And I have two questions:
1) Why there were accumulated so many files?
2) What I should do, when I have I buffer on the disk that is bigger, than buffer.max_size?

Thanks.

**Comment by CasparChou** [comment]
something new just found

these no been exceed max_size, but stuck and not sinking for hours

![image](https://user-images.githubusercontent.com/10682620/111284911-9131ee00-867b-11eb-92b1-14976157f55f.png)

vector connection abnormal (some agent keep the connection with aggregator, but aggregator seems not keep the conn)
![image](https://user-images.githubusercontent.com/10682620/111285127-c5a5aa00-867b-11eb-8877-f060d1016cbc.png)

**Comment by ktff** [comment]
@mikhno-s  thanks for reporting it, it helps.

Regarding 
>1. Why there were accumulated so many files?

#6602 is the issue for that problem so if you can contribute some information please do. Most notably if you are running Vector in Docker.

> 2. What I should do, when I have I buffer on the disk that is bigger, than buffer.max_size?

There shouldn't be a large difference between configured `max_size` and actual size so that's a bug to fix.

**Comment by ktff** [comment]
@mikhno-s 
>  Why there were accumulated so many files?
>  What I should do, when I have I buffer on the disk that is bigger, than buffer.max_size?
> I have 3GB mem limit and OOM kills vector in 30-60s after start.

fixes for this three has landed on master and is currently on nightly, while `v0.14` should be released soon . Could you check it out? 

@CasparChou I think the bug you reported is also fixed by a combination of changes, but I'm not so sure, so could you also check if it's still present?

**Comment by ktff** [comment]
This is fixed in `v0.14` but we can reopen it if it turns out that there are more causes.
